Enhance next gen event budget tool

outdoor2kode opened the pull request:
<!-- Start by describing the changes made in this Pull Request, and the reason for such changes. -->

<!-- Reference any related issues or PRs here. Each issue needs the "fixes" keyword if the PR fixes more than one thing. -->
See #976

This PR replaced, removed, and updated certain fields as requested in the ticket.
The modifications for `Days` and `Hours` will be in a separate PR, as they will involve more extensive changes to the logic than expected and this PR already has enough changes.

<!-- Don't forget to update the title with something descriptive. -->

### Screenshots

| Central Network | Event Network |
| -----------------|----------------|
| ![image](https://github.com/WordPress/wordcamp.org/assets/18050944/9814d522-5c21-4069-aa3a-16824fbe1740) | <img width="1120" alt="image" src="https://github.com/WordPress/wordcamp.org/assets/18050944/f06369d4-c4f0-45e5-8b7b-d3b52a011360"> ![image](https://github.com/WordPress/wordcamp.org/assets/18050944/7851c680-4447-4c9f-a309-92e37b07072d) |
| ![image](https://github.com/WordPress/wordcamp.org/assets/18050944/85317325-4313-4f84-9641-a0d49940eb77) | ![image](https://github.com/WordPress/wordcamp.org/assets/18050944/92943dcb-4e8e-404f-a090-8b39c47b87f9) |
| ![image](https://github.com/WordPress/wordcamp.org/assets/18050944/b481f417-114b-4f13-9efe-d5eff61e246e) | ![image](https://github.com/WordPress/wordcamp.org/assets/18050944/3ab85225-f99d-4fc5-be91-1b9f754086cc) |

 <!-- If your change has a visual component, add a screenshot here. A "before" screenshot would also be helpful. --> 

### How to test the changes in this Pull Request:

1. Create a new (a) og WordCamp site (b) next-gen event site through `WordCamps` >> `Add new`
2. In the dashboard of both sites, go to Budgets.
3. Check if the content is the same as the screenshots above demonstrate as well as requested in the ticket.

outdoor2kode committed:
Payments: Update the text Speakers to Facilitators for Event Data

outdoor2kode committed:
Payments: Remove Speaker Event, After Party, and Camera Shipping as expense categories

outdoor2kode committed:
Payments: add new field to the event data

WP Expertise Level, Focused Activity, Job Status, Identity Based, Content Topic Focused, and Other are added.

outdoor2kode committed:
Correct the ordering

outdoor2kode committed:
Payment: make dropdown as value for new fields

outdoor2kode committed:
Fix select style

outdoor2kode committed:
Payments: fill up new fields dropdown with options

outdoor2kode committed:
Add pointer cursor to select

outdoor2kode committed:
Payments: Filter out tracks when is_next_gen_wordcamp

outdoor2kode committed:
Fix the testing error

tellyworth commented on public_html/wp-content/plugins/wordcamp-payments/includes/budget-tool.php:
```diff
@@ -191,6 +191,14 @@ public static function enqueue_scripts() {
+				'wcb-budget-tool',
+				'networkStatus',
+				array(
+					'isNextGenWordCamp' => is_next_gen_wordcamp(),
```
How certain are we that there will only ever be two types of events (nextgen and OG)?

If it's likely there could be a third type in future then perhaps something other than a binary is/not distinction is worthwhile.

outdoor2kode replied:
Good question. 🤔  I'm not entirely certain, but adapting the function to accept a different type of event also sounds legit and good now.

outdoor2kode replied:
Updated in 294127b124dc48fcdf593169ca0da166922206d5

tellyworth commented on public_html/wp-content/plugins/wordcamp-payments/includes/budget-tool.php:
```diff
@@ -288,6 +319,25 @@ private static function _get_default_budget() {
+		if ( is_next_gen_wordcamp() ) {
+			// $extra_budget_for_next_gen should be positioned after the 'days' element for proper ordering.
+			$insert_position = array_search( 'days', array_column( $default_budget, 'name' ), true ) + 1;
+			array_splice( $default_budget, $insert_position, 0, $extra_budget_for_next_gen );
```
`array_splice()` with an offset of zero is a clever way of extending an array. I typically use a construction like `$default = array_merge( $default, $extra)`. I'd be curious to check out the performance differences; certainly array_splice could be more memory efficient.

StevenDufresne replied:
`array_merge` is definitely more readable.

outdoor2kode replied:
Actually, the offset is `$insert_position` here, not zero. 
In this case, `$extra_budget_for_next_gen` needs to be inserted immediately after `days`. If using `$default = array_merge($default, $extra);`, $extra would be appended to the end of the $default array.
I may have missed something or misunderstood your point, it'd be nice if you can correct me if that's the case, thanks.

tellyworth commented on public_html/wp-content/plugins/wordcamp-payments/includes/budget-tool.php:
```diff
@@ -254,12 +262,35 @@ private static function _get_budget() {
 
+	/**
+	 * Helper function to filter out the budget items.
+	 */
```
I'm a bit puzzled by the need for this. Why use dynamic filtering rather than just leave out the unneeded items from the declaration in `_get_default_budget()`? I'm sure there's a reason but it's not obvious.

outdoor2kode replied:
Because for og wordcamps, the items are still needed. If leaving out those items directly from the declaration, on og wordcamp sites, they won't show up.

---

Also, since the `_get_default_budget` is getting bigger, it seems to me that it's not easy to read and keep up with. And I'm guessing we'll want to pull out more items from the event network down the line, so I abstract this logic into a separate function. This would facilitate the addition or removal of items as needed. If we were to directly integrate this logic into `_get_default_budget`, I feel it would make the function a bit lengthy, especially as more changes are added down the road."


If we don't add an extra filtering function, we'd just be incorporating one of the following into `_get_default_budget`. It looks okay for now, but I'm thinking it might get a bit lengthy if we add more conditions down the line. What do you think?
```
$default_budget = array_filter($default_budget, function($item) {
    return !($item['name'] === 'tracks' || (isset($item['category']) && $item['category'] === 'speaker-event'));
});
```

```
$default_budget = array_filter($default_budget, function($item) {
    if (isset($item['name']) && $item['name'] === 'tracks') {
        return false;
    }
    
    if (isset($item['category']) && $item['category'] === 'speaker-event') {
        return false;
    }

    return true;
});
```

outdoor2kode replied:
~Or how do you feel about splitting the default budget retrieval for the next-gen network, like what's done in this [commit](https://github.com/WordPress/wordcamp.org/pull/998/commits/7c5dcc6a7687799f953318b15d0ebc68de380bbd)?~

the function was removed as the original requirements have changed.

tellyworth reviewed (approved):
No blockers here, and it looks fine when I tested per the test steps.

StevenDufresne commented on public_html/wp-content/plugins/wordcamp-payments/bootstrap.php:
```diff
@@ -23,6 +23,15 @@
+	 *
+	 * @return bool Returns true if the current site is a "next generation" WordCamp, otherwise false.
+	 */
+	function is_next_gen_wordcamp() {
```
As noted somewhere else in a different comment. Let's move this somewhere else. 

I think:https://github.com/WordPress/wordcamp.org/blob/production/public_html/wp-content/mu-plugins/3-helpers-misc.php is a good candidate.

outdoor2kode replied:
Updated in d870b2f050b8639fb256e27197a0f2d09506207f

outdoor2kode committed:
Separate retrieval for next-gen get_default_budget

outdoor2kode committed:
move is_next_gen_wordcamp to higher level

outdoor2kode committed:
Adapt is_next_gen_wordcamp to a more general function

It used to only distinguish between next-gen WordCamp or not. Now, it can verify any type of WordCamp.

outdoor2kode committed:
Revert: Payments: add new field to the event data

WP Expertise Level, Focused Activity, Job Status, Identity Based, Content Topic Focused, and Other are removed.

outdoor2kode committed:
Payments: Add field 'Format'

outdoor2kode committed:
Revert: Separate retrieval for next-gen get_default_budget
